src/lib.rs

```diff
@@ -126,8 +139,90 @@ pub fn select_coin_knapsack(
 fn knap_sack(
     adjusted_target: u64,
     smaller_coins: &[(usize, OutputGroup)],
+    inputs: &[OutputGroup],
+    options: CoinSelectionOpt,
 ) -> Result<SelectionOutput, SelectionError> {
-    unimplemented!()
+    let mut target_reached = false;
+    let mut selected_inputs: HashSet<(usize)> = HashSet::new();
+    let mut accumulated_value: u64 = 0;
+    let mut accumulated_weight: u32 = 0;
+    let mut best_set: HashSet<(usize)> = HashSet::new();
+    // Assigning infinity to beginwith
+    let mut best_set_value: u64 = u64::MAX;
+    let mut best_set_weight: u32 = 0;
+    let mut rng = thread_rng();
+    for i in 1..=1000 {
+        for pass in 1..=2 {
+            if !target_reached {
```

Comment:
There is a case when, accumulated_value>=adjusted_target, but !(accumulated_value<best_set_value). 
` } else if accumulated_value >= adjusted_target {
                            target_reached = true;
                            if accumulated_value < best_set_value {`
In that case, the algorithm should not try to find another input 
`if !target_reached {
                for &(index, u) in smaller_coins {
                    //Simulate a coin toss
                    let toss_result: bool = rng.gen_bool(0.5);`
The if !target_reached { condition is used to handle the above mentioned condition. Is there another way to handle that condition ?

Reply:
The `targetReached` variable in the algorithm 3 of the paper is a mistake.
For knapsack you work on list of smaller coins in descending order.
In case you have sufficient funds there are 2 scenarios,

1. You find exact match and return
2. You have more than the target and can potentially find smaller targets by deselecting the current one.

So you are not meant to terminate when you have found the target but to look for smaller coins to find close enough
match.
But the way the algorithm is implemented in the paper will return once the target is reached because that variable is
never set to false. At line 15 of the algorithm it is set to true and no more iterations will happen, and clearly this is not the
intended way. That's why you are setting the variable to false after deselecting the current selection once the target is reached.

So my question is, what purpose does this variable serve? You can deduce whether the target is reached or not
just by examining the `best_set_value` variable.

To your question of,
>>> There is a case when, accumulated_value>=adjusted_target, but !(accumulated_value<best_set_value).

In that case you are deselecting the current selection, and in the next iterations trying to select smaller coins which might give a better selection.

Reply:
Better thing to do is to move the deselection out of the
`line 189: if accumulated_value < best_set_value` condition,
because if the current selection is greater than the target but not smaller than the best_value
there is still a possibility with smaller coins to the right.
The computation should not stop there